### PR TITLE
[SPARK-39375][CONNECT][FOLLOW-UP] Refactor Read to UnresolvedRelation

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Refactor proto from `Read` to `UnresolvedRelation`.

### Why are the changes needed?

UnresolvedRelation has its own proto. In the future we will add a proto for DataSource. We can decouple these two plans in connect.

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @cloud-fan

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -31,7 +31,7 @@ option java_package = "org.apache.spark.connect.proto";
 message Relation {
   RelationCommon common = 1;
   oneof rel_type {
-    Read read = 2;
+    UnresolvedRelation unresolved_relation = 2;
```
Comment (grundprinzip) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: why?

**Comment by grundprinzip** [comment]
I think this change is unnecessary.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: Per the review suggestions on https://github.com/apache/spark/pull/38086, we plan to decouple unresolved relation and data source (e.g. in different messages). cc @cloud-fan 

Data source needs its own plan with its own fields. In that case current `Read` will only be unresolved relation.

Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: I found refactoring in https://github.com/apache/spark/pull/38086 makes that PR too complicated so I split this part out as a separate PR.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 2]
```diff
@@ -60,16 +60,9 @@ message SQL {
   string query = 1;
 }
 
-// Relation that reads from a file / table or other data source. Does not have additional
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:63]: @grundprinzip It's hard to unify this. Spark has 2 different code paths to resolve a table via name, and resolve a data source via format name, options, etc. This is kind of merging two query plans into one when they have no common fields. It's better to split.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: I don't agree with the change. a read is a sink the same way you use it in the from statement. 

What's wrong with modeling the different read types below Read? 

Conceptually, the hierarchical modeling of Read -> Table and Read-> data source makes a lot of sense.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: I left a comment in https://github.com/apache/spark/pull/38193#discussion_r991722481 but let me also reply here.

Yes, we want the query plan proto definition to follow the basics for the relational algebra / SQL language. However, data source reading is not really SQL, but more of "configuration". To read a data source in SQL, we need to do
```
CRETE TABLE t USING my_source OPTIONS (...)
SELECT * FROM t
```
We must register the data source as a table and then read it. We can't read a data source directly in SQL.

DF API provides a way to read a data source directly: `spark.read.format("my_source").option(...).load()`. This means, we can't say `a read is a sink the same way you use it in the from statement`. Reading a data source directly needs a dedicated query plan.

Comment (hvanhovell) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: I am not sure I follow why we need to bifurcate into different types of reads in the proto. The whole point of using parsed plans is to have a high level (non-ambiguous) description of the operation, not an implementation, the latter is for Spark to figure out. We just need to describe whatever it takes what is needed to read the data. If it is a table scan then it is easy, you only need a name, otherwise you may have to add a bit more information.

BTW nitpicking on naming. Operations should almost always have a verb as its name (scan, project, filter, aggregate, generate, ...) if it does not, you probably should rethink the name. `UNRESOLVED RELATION` is the thing you want to scan, and not the operation.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: @hvanhovell , yes, we are talking about how to describe the scan operation. IIUC there are two options on the table:
1. Have a single `Read` operation, with many fields: `tableName`, `userSpecifiedSchema`, `dataSourceName`, `options`. If we are scanning a table, only `tableName` is set, if we are scanning a data source, all fields except `tableName` are set.
2. Have a `UnresolvedRelation` and `UnresolvedDataSource` (we can pick better names, this is just to match catalyst). `UnresolvedRelation` only have a `tableName` field, `UnresolvedDataSource` has many fields: `userSpecifiedSchema`, `dataSourceName`, `options`.

I'm not familiar wth the protobuf design philosophy. Does it prefer option 1 or 2?

Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: Maybe we're getting closer to it. Yes, proto has proper Union type support. So what the `Read` relation should look like is the following:

```
message Read {

  oneof read_type {
      UnresolvedRelation unresolved_relation = 1;
      UnresolvedDataSource unresolved_data_source =2;
  }

  message UnresolvedRelation {
  }

  message UnresolvedDataSource {
  }

}
```

The generated code has the right behavior.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 3]
```diff
@@ -60,16 +60,9 @@ message SQL {
...
-// inputs.
-message Read {
-  oneof read_type {
-    NamedTable named_table = 1;
```
Comment (grundprinzip) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:67]: @cloud-fan, please see here, the Read had already had a union type and this just needs to be extended with an additional DataSource type.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:67]: I see, yea we can add a new `DataSource` type here, in addition to the `NamedTable`. This seems like we give them a parent class `Read`. Any benefits of doing so compared to just flattening it? Please bear with me as I'm not very familiar with protobuf...

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:34]: I was thinking something similar: 

It is a wrong pattern that a message contains many fields while a subset of the fields that belong to one type and others that belong to different types, which is option 1 that Wenchen mentioned.

It's good that we keep all relevant fields into one message. If there is more than one message but we want to co-locate it then there is a type to differentiate those, which is Wenchen's option 2 and Martin's example, if I am not mis-read it.

I think we are sort of having the consensus on the rule of when to split messages. 


However the rule of when to co-locate messages here is still not clear on, for example, why we don't choose to define
```
message ExecuteWithSingleInput {
 oneof node_type {
     Project,
     Filter
 } 
 
 message Project {
 
 }
 
 message Filter {

 }
}
```

Basically the rule to split messages is clear, but the rule to co-locate message is not. We can co-locate `Project` and `Filter` as they both have single relation input. But we have chosen not to in the proto. The naming in my example is bad and I am pretty we can have a better verb to describe it.

To me the sharing semantic of a `unresolved_relation` and `data_source` is even less than `project` and `filter`. In Apache Calcite, it can merge `project` and `filter` and call it a `Calc`
